applications/OptimizationApplication/custom_utilities/filtering/explicit_vertex_morphing_filter.cpp

```diff
@@ -0,0 +1,311 @@
+        << "Uninitialized container expression given. "
+        << rContainerExpression;
+
+    KRATOS_ERROR_IF_NOT(&rContainerExpression.GetModelPart() == &mrModelPart)
+        << "Filter radius container expression model part and filter model part mismatch."
+        << "\n\tFilter = " << *this
+        << "\n\tContainerExpression = " << rContainerExpression;
+
+    const IndexType stride = rContainerExpression.GetItemComponentCount();
+    const auto& r_origin_expression = rContainerExpression.GetExpression();
+    const auto& r_container = rContainerExpression.GetContainer();
+    const auto& r_filter_radius_expression = mpFilterRadiusContainer->GetExpression();
+
+    ContainerExpression<TContainerType> result(*rContainerExpression.pGetModelPart());
+    auto p_expression = LiteralFlatExpression<double>::Create(result.GetContainer().size(), rContainerExpression.GetItemShape());
+    result.SetExpression(p_expression);
+
+    struct TLS
+    {
+        explicit TLS(const IndexType MaxNumberOfNeighbors)
+        {
+            mNeighbourEntityPoints.resize(MaxNumberOfNeighbors);
+            mResultingSquaredDistances.resize(MaxNumberOfNeighbors);
+        }
+
+        EntityPointVector mNeighbourEntityPoints;
+        std::vector<double> mResultingSquaredDistances;
+    };
+
+    IndexPartition<IndexType>(r_container.size()).for_each(TLS(mMaxNumberOfNeighbors), [&](const IndexType Index, TLS& rTLS){
+        const double vertex_morphing_radius = r_filter_radius_expression.Evaluate(Index, Index, 0);
+
+        EntityPoint<EntityType> entity_point(*(r_container.begin() + Index), Index);
+        const auto number_of_neighbors = mpSearchTree->SearchInRadius(
+                                            entity_point,
+                                            vertex_morphing_radius,
+                                            rTLS.mNeighbourEntityPoints.begin(),
+                                            rTLS.mResultingSquaredDistances.begin(),
+                                            mMaxNumberOfNeighbors);
+
+        std::vector<double> list_of_weights(number_of_neighbors, 0.0);
+        double sum_of_weights = 0.0;
+        ExplicitVertexMorphingFilterHelperUtilities::ComputeWeightForAllNeighbors(
+            sum_of_weights, list_of_weights, *mpKernelFunction, vertex_morphing_radius,
+            entity_point, rTLS.mNeighbourEntityPoints, number_of_neighbors);
+
+        const IndexType current_data_begin = Index * stride;
+
+        for (IndexType j = 0; j < stride; ++j) {
+            double& current_index_value = *(p_expression->begin() + current_data_begin + j);
+            current_index_value = 0.0;
+
+            for(IndexType neighbour_index = 0 ; neighbour_index < number_of_neighbors; ++neighbour_index) {
+                const IndexType neighbour_id = rTLS.mNeighbourEntityPoints[neighbour_index]->Id();
+                const double weight = list_of_weights[neighbour_index] / sum_of_weights;
```

Comment:
You need to include the area for surface field filtering and the volume for the volume field filtering. So the weight of each node is multiplied by area/volume and also considered in the sum of weights. In this way, we can make sure that the filtering of a discrete field is mesh independent.

Reply:
Please also implement FilterIntegratedField which only considers integration in the sum of the weights because the given field is already integrated, like the discrete sensitivity of structures and fluids.

Reply:
I would suggest to only have FilterField and FilterIntegratedField, which do the job for forward and backward mapping.

Reply:
Added area contribution and `FilterIntegratedField`